docs/spec/spec.md

```diff
@@ -0,0 +1,1821 @@
+default resource can act as a common destination where the unmatched requests (either HTTP method or resource path) may 
+get dispatched.
+
+```ballerina
+resource function 'default [string... s]() {
+
+}
+```
+
+#### 2.3.4. Signature parameters
+The resource function can have the following parameters in the signature. There are no any mandatory params or any 
+particular order. But it’s a good practice to keep the optional param at the end.
+
+```ballerina
+resource function XXX NAME_TEMPLATE ([http:Caller hc], [http:Request req], (anydata queryParam)?, 
+    (@http:Payload anydata payload)?, (@http:Header string header)?, (http:Header headers)? ) {
+        
+}
+```
+
+However, the first choice should be to use signature params and use returns. Avoid caller unless you have specific 
+requirement. Also use data binding, header params and resource returns to write smaller code with more readability.
+
+##### 2.3.4.1. http:Caller
+
+The caller client object represents the endpoint which initiates the request. Once the request is processed, the 
+corresponding response is sent back using the remote functions which are associated with the caller object. 
+In addition to that, the caller has certain meta information related to remote and local host such as IP address,
+protocol. This parameter is not compulsory and not ordered.
+
+
+The callerInfo annotation associated with the `Caller` is to denote the response type.
+It will ensure that the resource function responds with the right type and provides static type information about 
+the response type that can be used to generate OpenAPI.
+
+The default type is the `http:Response`. Other than that, caller remote functions will accept following types as the 
+outbound response payload. Internally an `http:Response` is created including the given payload value
+
+```ballerina
+string|xml|json|byte[]|int|float|decimal|boolean|map<json>|table<map<json>>|(map<json>|table<map<json>>)[]|
+mime:Entity[]|stream<byte[], io:Error?>|()
+```
+
+Based on the payload types respective header value is added as the `Content-type` of the `http:Response`.
+
+ Type|Content Type
+---|---
+() | -
+string | text/plain
+xml | application/xml
+byte[], stream<byte[], io:Error?> | application/octet-stream
+int, float, decimal, boolean | application/json
+map\<json\>, table<map\<json\>>, map\<json\>[], table<map\<json\>>)[] | application/json
+
+The HTTP compiler extension checks the argument of the `respond()` method if the matching payload type is passed as
+denoted in the CallerInfo annotation
+
+```ballerina
+resource function post foo(@http:CallerInfo {respondType:Person}  http:Caller hc) {
+    Person p = {};
+    hc->respond(p);
+}
+```
+##### 2.3.4.2. http:Request
+
+The `http:Request` represents the message which came along over the network that modelled as an object along with 
+headers and payload. Listener passes it to the resource function as an argument to be accessed by the user based on 
```

Comment:
I think the first sentence can be changed to something like the below.

The `http:Request` represents the request which is sent and received over the network which includes headers and the entity body.